pkg/controller/jobs/pod/pod_webhook.go

```diff
@@ -0,0 +1,155 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+)
+
+const (
+	ManagedLabelKey   = "kueue.x-k8s.io/managed"
+	ManagedLabelValue = "true"
+	PodFinalizer      = ManagedLabelKey
+)
+
+type PodWebhook struct {
+	client                     client.Client
+	manageJobsWithoutQueueName bool
+	namespaceSelector          *metav1.LabelSelector
+	podSelector                *metav1.LabelSelector
+}
+
+// SetupWebhook configures the webhook for pods.
+func SetupWebhook(mgr ctrl.Manager, opts ...jobframework.Option) error {
+	options := jobframework.DefaultOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+	wh := &PodWebhook{
+		client:                     mgr.GetClient(),
+		manageJobsWithoutQueueName: options.ManageJobsWithoutQueueName,
+		namespaceSelector:          options.PodNamespaceSelector,
+		podSelector:                options.PodSelector,
+	}
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&corev1.Pod{}).
+		WithDefaulter(wh).
+		WithValidator(wh).
+		Complete()
+}
+
+func fromObject(o runtime.Object) *Pod {
+	return (*Pod)(o.(*corev1.Pod))
+}
+
+// +kubebuilder:webhook:path=/mutate--v1-pod,mutating=true,failurePolicy=ignore,sideEffects=None,groups="",resources=pods,verbs=create,versions=v1,name=mpod.kb.io,admissionReviewVersions=v1
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+
+var _ webhook.CustomDefaulter = &PodWebhook{}
+
+func (w *PodWebhook) Default(ctx context.Context, obj runtime.Object) error {
+	pod := fromObject(obj)
+	log := ctrl.LoggerFrom(ctx).WithName("pod-webhook").WithValues("pod", klog.KObj(pod))
+	log.V(5).Info("Applying defaults")
+
+	if IsPodOwnerManagedByQueue(pod) {
+		log.V(5).Info("Pod owner is managed by queue, skipping")
+		return nil
+	}
+
+	// Check for pod label selector match
+	podSelector, err := metav1.LabelSelectorAsSelector(w.podSelector)
+	if err != nil {
+		return fmt.Errorf("failed to parse pod selector: %w", err)
+	}
+	if !podSelector.Matches(labels.Set(pod.GetLabels())) {
+		return nil
+	}
+
+	// Get pod namespace and check for namespace label selector match
+	ns := corev1.Namespace{}
+	err = w.client.Get(ctx, client.ObjectKey{Name: pod.GetNamespace()}, &ns)
+	if err != nil {
+		return fmt.Errorf("failed to run mutating webhook on pod %s, error while getting namespace: %w",
+			pod.GetName(),
+			err,
+		)
+	}
+	log.V(5).Info("Found pod namespace", "Namespace.Name", ns.GetName())
+	nsSelector, err := metav1.LabelSelectorAsSelector(w.namespaceSelector)
+	if err != nil {
+		return fmt.Errorf("failed to parse namespace selector: %w", err)
+	}
+	if !nsSelector.Matches(labels.Set(ns.GetLabels())) {
+		return nil
+	}
+
+	if jobframework.QueueName(pod) != "" || w.manageJobsWithoutQueueName {
+		controllerutil.AddFinalizer(obj.(*corev1.Pod), PodFinalizer)
```

Comment:
```suggestion
		controllerutil.AddFinalizer(pod, PodFinalizer)
```

Reply:
The underlying type of `pod` here is: `corev1.Pod`. `AddFinalizer` expects `*corev1.Pod`. 

Is there a way to change the underlying type using the Pod instance?

Reply:
what about `pod.Object()`?

I thought `pod` itself could work too, since `AddFinalizer` accepts an interface that either `pod` or `pod.Object()` implement.

Reply:
`pod.Object()` works fine: https://github.com/kubernetes-sigs/kueue/blob/1c7a932045462bf96865a33213d228549dd02634/pkg/controller/jobs/pod/pod_webhook.go#L118

`AddFinalizer` accepts `client.Object` interface with `DeepCopyObject` method. And `DeepCopyObject` for `corev1.Pod` has a pointer receiver.